src/blog/sparkplug.md

```diff
@@ -0,0 +1,188 @@
+
+The second trick is that Sparkplug doesn't generate any intermediate representation (IR) like most compilers do. Instead, Sparkplug compiles directly to machine code in a single linear pass over the bytecode, emitting code that matches the execution of that bytecode. In fact, the entire compiler is a [switch statement](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/baseline/baseline-compiler.cc;l=465;drc=55cbb2ce3be503d9096688b72d5af0e40a9e598b) inside a [for loop](https://source.chromium.org/chromium/chromium/src/+/main:v8/src/baseline/baseline-compiler.cc;l=290;drc=9013bf7765d7febaa58224542782307fa952ac14), dispatching to fixed per-bytecode machine code generation functions.
+
+```cpp
+// The Sparkplug compiler (abridged).
+for (; !iterator.done(); iterator.Advance()) {
+  VisitSingleBytecode();
+}
+```
+
+The lack of IR means that the compiler has limited optimisation opportunity, beyond very local peephole optimisations. It also means that we have to port the entire implementation separately to each architecture we support, since there's no intermediate architecture-independent stage. But, it turns out that neither of these is a problem: a fast compiler is a simple compiler, so the code is pretty easy to port; and Sparkplug doesn't need to do heavy optimisation, since we have a great optimising compiler later on in the pipeline anyway.
+
+::: note
+Technically, we currently do two passes over the bytecode — one to discover loops, and a second one to generate the actual code. We're planning on getting rid of the first one eventually though.
+:::
+
+## Interpreter-compatible frames
+
+Adding a new compiler to an existing mature JavaScript VM is a daunting task. There's all sorts of things you have to support beyond just standard execution; V8 has a debugger, a stack-walking CPU profiler, there's stack traces for exceptions, integration into the tier-up, on-stack replacement to optimized code for hot loops… it's a lot.
+
+Sparkplug does a neat sleight-of-hand that simplifies most of these problems away, which is that it maintains "interpreter-compatible stack frames".
+
+Let's rewind a bit. Stack frames are how code execution stores function state; whenever you call a new function, it creates a new stack frame for that function's local variables. A stack frame is defined by a frame pointer (marking its start) and a stack pointer (marking its end):
+
+![A stack frame, with stack and frame pointers](/_svg/sparkplug/basic-frame.svg)
+
+::: note
+<!-- markdownlint-capture -->
+<!-- markdownlint-disable no-inline-html -->
+At this point, roughly half of you will be screaming, saying "this diagram doesn't make sense, stacks obviously grow in the opposite direction!". Fear not, I made a button for you: <button id="flipStacksButton">I think stacks grow upwards</button>
+<script>
+  const flipStacksButton = document.getElementById("flipStacksButton");
+  let stacksAreFlipped;
+  if (localStorage['stacksAreFlipped'] !== undefined) {
```

Comment:
Let's get rid of `localStorage` as discussed locally (to avoid polluting local storage for visitors forever).

Reply:
Done